subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,733 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_TLS_DEBUG)
+#define SYS_LOG_DOMAIN "net/tls"
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <stdio.h>
+#include <sys/param.h>
+
+#include "net_private.h"
+#include "net_tls_internal.h"
+#include "tcp_internal.h"
+
+#define TIMEOUT_TLS_RX_MS       100
+#define TIMEOUT_TLS_TX_MS       100
+
+#if defined(CONFIG_NET_PRECONFIGURE_TLS_CREDENTIALS)
+#include CONFIG_NET_TLS_CREDENTIALS_FILE
+#endif
+
+static mbedtls_ctr_drbg_context tls_ctr_drbg;
+
+struct net_tls_credential {
+	enum net_tls_credential_type type;
+	sec_tag_t tag;
+	const void *buf;
+	size_t len;
+};
+
+static struct net_tls_credential credentials[CONFIG_NET_MAX_CREDENTIALS_NUMBER];
+
+static struct net_tls tls_context[CONFIG_NET_TLS_MAX_CONTEXT];
+
+#if defined(MBEDTLS_DEBUG_C) && defined(CONFIG_NET_TLS_DEBUG)
+static void my_debug(void *ctx, int level,
+		     const char *file, int line, const char *str)
+{
+	const char *p, *basename;
+	int len;
+
+	ARG_UNUSED(ctx);
+
+	/* Extract basename from file */
+	for (p = basename = file; *p != '\0'; p++) {
+		if (*p == '/' || *p == '\\') {
+			basename = p + 1;
+		}
+
+	}
+
+	/* Avoid printing double newlines */
+	len = strlen(str);
+	if (str[len - 1] == '\n') {
+		((char *)str)[len - 1] = '\0';
+	}
+
+	NET_DBG("%s:%04d: |%d| %s", basename, line, level, str);
+}
+#endif /* defined(MBEDTLS_DEBUG_C) && defined(CONFIG_NET_TLS_DEBUG) */
+
+static int tls_entropy_func(void *ctx, unsigned char *buf, size_t len)
+{
+	/* TODO Provide a decent entropy source. */
+
+	ARG_UNUSED(ctx);
+
+	size_t i = len / 4;
+	u32_t val;
+
+	while (i--) {
+		val = sys_rand32_get();
+		UNALIGNED_PUT(val, (u32_t *)buf);
+		buf += 4;
+	}
+
+	i = len & 0x3;
+	val = sys_rand32_get();
+	while (i--) {
+		*buf++ = val;
+		val >>= 8;
+	}
+
+	return 0;
+}
+
+static struct net_tls_credential *unused_credential_get(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
```

Comment:
All the "NET_TLS || NET_DTLS subsystem" should be marked as thread-unsafe with this fine design.

Reply:
> All the "NET_TLS || NET_DTLS subsystem" should be marked as thread-unsafe with this fine design.

We've had this discussion before, but I'll repeat that it is not safe to run the IP stack in multipile threads. Marking only this instance as thread unsafe will make developers believe that the rest are safe, which is not the case. Let's start looking at the IP stack thread safetiness once we have enabled sockets, updated network buffers and memory protection.

Reply:
> We've had this discussion before, but I'll repeat that it is not safe to run the IP stack in multipile threads. Marking only this instance as thread unsafe

Well, the fact we goofed and produced a broken IP stack doesn't mean that we can shove more broken code into it. Leaving fixing all for "later" means that it will be very hard and time-consuming to fix it. (https://github.com/zephyrproject-rtos/zephyr/issues/729 is one year old and it was only now confirmed that the intuition behind it was right.)

There's nothing wrong with one thread creating a TLS connection while another doing the same, and the current code will have issues in this case - which will take some time to surface for people, and then very hard to reproduce. So, it rather should be fixed now (and fairly speaking that's what imply as an alternative of documenting non-thread-safety).